aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains test case classes based on pytest for AEA skill testing."""
+
+import asyncio
+from pathlib import Path
+from queue import Queue
+from types import SimpleNamespace
+from typing import Optional, Union, cast
+
+from aea.context.base import AgentContext
+from aea.identity.base import Identity
+from aea.multiplexer import AsyncMultiplexer, Multiplexer, OutBox
+from aea.protocols.base import Message
+from aea.skills.base import Skill
+from aea.skills.tasks import TaskManager
+
+
+class BaseSkillTestCase:
+    """A class to test a skill."""
+
+    path_to_skill: Union[Path, str] = Path(".")
+    _skill: Skill
+    _multiplexer: AsyncMultiplexer
+    _outbox: OutBox
+
+    @property
+    def skill(self) -> Skill:
+        """Get the skill."""
+        try:
+            value = self._skill
+        except AttributeError:
+            raise ValueError("Ensure skill is set during setup_class.")
+        return value
+
+    def get_quantity_in_outbox(self) -> int:
+        """Get the quantity of envelopes in the outbox."""
+        return self._outbox._multiplexer.out_queue.qsize()
+
+    def get_message_from_outbox(self) -> Optional[Message]:
+        """Get message from outbox."""
+        if self._outbox.empty():
+            return None
+        envelope = self._multiplexer.out_queue.get_nowait()
+        return envelope.message
+
+    @classmethod
+    def setup(cls) -> None:
+        """Set up the skill test case."""
+        identity = Identity("test_agent_name", "test_agent_address")
+
+        cls._multiplexer = AsyncMultiplexer()
+        cls._multiplexer._out_queue = asyncio.Queue()
```

Comment:
Multiplexer can be removed here if we instantiate the out_queue in the outbox..can be trivially done but should be a separate PR (i.e. on connect call multiplexer calls outbox to set out_queue)